pkg/scheduler/preemption/preemption.go

```diff
@@ -77,7 +87,32 @@ func (p *Preemptor) Do(ctx context.Context, wl workload.Info, assignment flavora
 	}
 	sort.Slice(candidates, candidatesOrdering(candidates, cq.Name, time.Now()))
 
-	targets := minimalPreemptions(&wl, assignment, snapshot, resPerFlv, candidates)
+	sameQueueCandidates := candidatesOnlyFromQueue(candidates, wl.ClusterQueue)
+	var targets []*workload.Info
+	triedBorrowing := false
+
+	// To avoid flapping, Kueue only allows preemption of workloads from the same
+	// queue if borrowing. Preemption of workloads from queues can happen only
+	// if not borrowing at the same time. Kueue prioritizes preemption of
+	// workloads from the other queues (that borrowed resources) first, before
+	// trying to preempt more own workloads and borrow at the same time.
+
+	if len(sameQueueCandidates) == len(candidates) {
+		// There is no risk of preemption of workloads from the other queue,
+		// so we can try borrowing.
+		triedBorrowing = true
+		targets = minimalPreemptions(&wl, assignment, snapshot, resPerFlv, candidates, true)
+	} else {
+		// There is a risk of preemption of workloads from the other queue in the
+		// cohort, proceeding without borrowing.
+		targets = minimalPreemptions(&wl, assignment, snapshot, resPerFlv, candidates, false)
+	}
+	if len(targets) == 0 && len(sameQueueCandidates) != len(candidates) && !triedBorrowing {
```

Comment:
This conditional can be inside the else above, instead of checking the lengths again. And we don't need the `triedBorrowing` variable.